### PR TITLE
Upcasting, scaling, masking and fused kernels to match Megatron-LM

### PR DESCRIPTION
**Pull request by jlamypoirier**
Run the following sequence (when enabled):
* attn_weights = (query x key) / sqrt(head_dim) / (layer_idx+1) [Single kernel, I think that's needed to stabilize fp16]
* Convert to fp32
* multiply by (layer_idx+1)
* Apply mask
* Softmax
* Convert to fp16

The whole thing runs in 2 kernels with jit fusion. Cuda time inference benchmark:
* Without jit: 793.142ms
* With jit: 751.869ms

Without upcast, the sequence is:
Run the following sequence (when enabled):
* attn_weights = (query x key) / sqrt(head_dim)
* Apply mask
* Softmax

Jit can reduce it to 2 kernels:
* Without jit: 745.253ms
* With jit: 751.365ms

Jit looks slower so I disabled it.

Overall, jit fusion allows fp32 softmax with minimal overhead (0.9% instead of 6.4%) and without the complicated custom kernels from Megatron.

### CODE REVIEW

File: src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py [hunk 1]
```diff
@@ -790,34 +809,28 @@ def forward(
             position_ids = torch.arange(past_length, input_shape[-1] + past_length, dtype=torch.long, device=device)
             position_ids = position_ids.unsqueeze(0).view(-1, input_shape[-1])
 
-        # GPTBigCodeAttention mask.
+        # Self-attention mask.
+        query_length = input_shape[-1]
+        key_length = past_length + query_length
+        self_attention_mask = self.bias[None, key_length - query_length : key_length, :key_length]
```
Comment (RaymondLi0) [inline src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:815]: Is it necessary to move the causal-masking part from the Attention module to the `GPTBigCodeModel`?

Comment (jlamypoirier) [reply src/transformers/models/gpt_bigcode/modeling_gpt_bigcode.py:815]: It's an extra optimization, so there is a single masking op instead of two and we avoid duplicating the mask generation.
